src/DIRAC/Resources/Computing/CloudComputingElement.py

```diff
@@ -0,0 +1,549 @@
+""" Cloud Computing Element
+
+This allows submission to cloud sites using libcloud (via the standard
+SiteDirector agent). The instances are contextualised using cloud-init.
+
+Running cloud VM instances containing pilots is very analogous to classic cloud
+jobs. There are however some things that work differently:
+
+- File I/O: A small amount of input may be transferred through the
+  instance metadata, but after that the VM is inaccessible.
+- Authentication: Most cloud endpoints use a password or API style credentials
+  rather than a grid style proxy based authentication. The pilot still requires
+  a suitable proxy, but this cannot be renewed via the cloud interface due to
+  the I/O limitations.
+- Pilot (VM) Tidy-up: Cloud providers will not remove stopped instances by
+  default.
+
+To avoid the proxy renewal limitations, an alternate pilot proxy is used within
+the instances. This can either be a longer version of the usual pilot proxy or
+a pilot proxy generated from another dedicated cert/user. The proxy contains
+the DIRAC group, but no VOMS (as this would likely expire too quickly).
+
+By default it is assumed that a generic CentOS7 base image is being used. This
+will be fully contextualised using cloud-init:
+
+- CVMFS & Singularity will be installed.
+- A dirac user will be created to run the jobs.
+- Pilot proxy and start-up scripts will be installed in /mnt.
+- The usual pilot script will be placed in the dirac home directory and
+  the start-up scripts are run (as the dirac user).
+- After the pilot terminates, the machine is stopped by calling halt.
+
+A partially or fully pre-configured image may be used instead and the
+cloud-init template can be customised as necessary for this or any use case.
+This is recommended on production systems to cut-down on the overhead when
+starting many new instances.
+
+The majority of cloud providers identify instances with some form of unique
+identifier (generally a UUID), this is used in the pilot references. Each
+instance can generally also have a "friendly name" associated with it, which
+may not be unique. We set the friendly name to match a string that can be
+pattern matched; this allows any stopped instances to be found & removed
+automatically without affecting other VMs potentially running as the same user.
+
+Instances that match the "friendly name" prefix and have been running above a
+maximum lifetime are assumed to be stuck or lost and will be removed. This is
+to ensure that instances don't reserve/consume resources indefinitely.
+
+Most cloud authentication systems require some form of static secret such as a
```

Comment:
Why do you introduce another configuration file type and not using a CFG file ? In the latter case you would get the "secrets" data simple for ceParameters.

Reply:
This is to make sure it stays secret (our older openstack stuff had a simpler version of the same construct in it). I know there is some way to hide sensitive data in the CS as it's used for the database passwords but I felt it far safer to keep these details completely separate.

Reply:
Well, this can be made in a completely separate cfg file, just that it is passed to the SiteDirector in the run script. So, it is safe but in the code it is all in the same configuration space. This is how we use multiple SSHComputingElements with login/password access, those settings are not mixed with etc/dirac.cfg but are kept in a separate file, possibly more protected if needed.